internal/app/cache/cache.go

```diff
@@ -1,19 +1,134 @@
 package cache
 
 import (
-	envoy_api_v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/groupcache/lru"
+
+	v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
 )
 
 type Cache interface {
-	// Exists returns true if the key exists in the cache, false otherwise.
-	Exists(key string) bool
-
 	// Fetch returns the cached response if it exists.
-	Fetch(key string) (*envoy_api_v2.DiscoveryResponse, error)
+	Fetch(key string) (*v2.DiscoveryResponse, error)
+
+	// SetResponse sets the cache response and returns the list of requests.
+	SetResponse(key string, resp v2.DiscoveryResponse) ([]*v2.DiscoveryRequest, error)
+
+	// AddRequest adds the request to the cache and returns whether a stream is open.
+	AddRequest(key string, req v2.DiscoveryRequest) (bool, error)
+}
+
+type cache struct {
+	cacheMu sync.RWMutex
+	cache   lru.Cache
+	ttl     time.Duration
+}
+
+type resource struct {
+	resp           *v2.DiscoveryResponse
+	requests       []*v2.DiscoveryRequest
+	streamOpen     bool
+	expirationTime time.Time
+}
 
-	// SetResponse sets the cache response and returns the list of open watches.
-	SetResponse(key string, resp envoy_api_v2.DiscoveryResponse) ([]*envoy_api_v2.DiscoveryRequest, error)
+// Callback function for each eviction. Receives the key and cache value when called.
+type onEvictFunc func(key lru.Key, value interface{})
+
+func NewCache(maxEntries int, onEvicted onEvictFunc, ttl time.Duration) Cache {
+	return &cache{
+		cache: lru.Cache{
+			// Max number of cache entries before an item is evicted. Zero means no limit.
+			MaxEntries: maxEntries,
+			// OnEvict is called for each eviction and closes the stream if a key is removed
+			// (e.g. expiry due to TTL, too many entries).
+			OnEvicted: onEvicted,
+		},
+		// Duration before which an item is evicted for expiring. Zero means no expiration time.
+		ttl: ttl,
+	}
+}
+
+func (c *cache) Fetch(key string) (*v2.DiscoveryResponse, error) {
+	c.cacheMu.Lock()
+	defer c.cacheMu.Unlock()
+	value, found := c.cache.Get(key)
+	if !found {
+		return nil, fmt.Errorf("no value found for key: %s", key)
+	}
+	resource, ok := value.(resource)
+	if !ok {
+		return nil, fmt.Errorf("unable to cast cache value to type resource for key: %s", key)
+	}
+	// Lazy eviction based on TTL occurs here. Fetch does not increase the lifespan of the key.
+	if resource.isExpired() {
+		c.cache.Remove(key)
```

Comment:
is onEvicted not implemented yet? Do we envision to implement it here?

Reply:
I left a TODO for onEvicted as 1) we are letting it be a parameter that is passed in and 2) for our purposes, onEvicted must guarantee the stream is closed, which will likely require more work outside the scope of this PR as per https://github.com/envoyproxy/xds-relay/pull/25#discussion_r393441180.

There's no need to explicitly call onEvicted since LRUCache invokes it upon [Remove](https://github.com/golang/groupcache/blob/8c9f03a8e57eb486e42badaed3fb287da51807ba/lru/lru.go#L111).

Reply:
makes sense. Is there a way to test that onEvicted was called?

Reply:
I asked in the slack thread - I am not sure how to test it was called.

Reply:
onEvicted() is a callback func though right? Are you envisioning the caller to be responsible for shutting down the stream? If we're shutting down the stream as a part of the cache implementation, I would just add a TODO. It's separate from onEvict, the callback.

Reply:
regarding testing that onEvicted was called: I passed in a test eviction callback function that panics to verify the expected behavior. LMK if it's sufficient - I can remove if it's too hacky, but for now I do not know of another way to test that a function with no return value was called.

regarding shutting down the stream: I had thought we would be using onEvicted as a mechanism for ensuring that anytime an item is removed from the cache, we guarantee that its stream is closed. In my previous PR onEvicted was made configurable so that people could override it with their own behavior, which made it the responsibility of the caller rather than part of the cache implementation. We could shut down the stream when we know an item is removed via TTL but not when the cache is full because LRUCache internally does that, so I think if we want to keep onEvicted as a parameter, it will be on the caller to implement.

Reply:
Given that there is a TODO to have the cache implementation guarantee that the stream is opened, I'm inclined to have it also be responsible for closing the stream (and therefore making onEvicted something we set internally and not be configurable - or providing a callback we can default to if the user doesn't provide one).

Reply:
> I passed in a test eviction callback function that panics to verify the expected behavior.

I don't think it's hacky, think it's a sensible approach.

> Given that there is a TODO to have the cache implementation guarantee that the stream is opened, I'm inclined to have it also be responsible for closing the stream (and therefore making onEvicted something we set internally and not be configurable - or providing a callback we can default to if the user doesn't provide one).

Yeah, it goes back to the two trains of thoughts we were discussing:
Whether we want
1. orchestrator -> (on evict) -> issue the disconnect with multiplexer,
or
2. cache -> directly issue the disconnect with multiplexer.
cache would still set the `streamOpen` var to false in both scenarios. In scenario 1, it'd be called prior to triggering the callback.

Personally I prefer the first approach as I don't think the cache should be aware of the multiplexer. But that creates some questions of whether the cache should be aware of a `streamOpen` variable all together or if it should be moved to the orchestrator component. I think this was something I could have fleshed out a bit better on the original design. I'm going to think more on this aspect, dropping this here for feedback from both.

Reply:
Makes sense - I've gone ahead and removed all references to streamOpen in the cache implementation.

Reply:
thanks, let me know when you push up the change and I'll take another pass.

Reply:
Whoops, I thought I had pushed my changes. It's updated and ready for a look now.